### ISSUE TITLE
Python3.5: Absolute "import from" of a module was not supported

### ISSUE THREAD

**Issue by deajan** [Nuitka/Nuitka#618]
Hello Kay,

I think I found a bug in the way Nuitka handles local scope namespaces.
Here's what I got so far

* Nuitka version, full Python version and Platform (Windows, OSX, Linux ...)

0.6.8rc2
Python: 3.7.6 (tags/v3.7.6:43364a7ae0, Dec 19 2019, 00:42:30) [MSC v.1916 64 bit (AMD64)]
Executable: c:\Python37-64\python.exe
OS: Windows
Arch: x86_64

* How did you install Nuitka and Python (pip, anaconda, deb, rpm, from source,
  what is a virtualenv ...), this is very important usually.

Installed via `python.exe -m pip install "https://github.com/Nuitka/Nuitka/archive/factory.zip`

I tried (for the first time) to compile a simple hello world project that uses FastAPI and uvicorn ASGI server.

Here's the SSCEE 
```
from fastapi import FastAPI
import uvicorn

app = FastAPI()


@app.get("/")
async def read_root():
    return {"Hello": "World"}


@app.get("/items/{item_id}")
async def read_item(item_id: int, q: str = None):
    return {"item_id": item_id, "q": q}

if __name__ == "__main__":
    uvicorn.run("fapi_test:app", host="127.0.0.1", port=5000, log_level="info")
```

Both nuitka 0.6.7 and 0.6.8rc2 crash with the following output:
```
C:\TEST\fapi\fapi>c:\Python37-64\python.exe -m nuitka --standalone fapi_test.py
Traceback (most recent call last):
  File "c:\Python37-64\lib\site-packages\nuitka\__main__.py", line 186, in <module>
    main()
  File "c:\Python37-64\lib\site-packages\nuitka\__main__.py", line 179, in main
    MainControl.main()
  File "c:\Python37-64\lib\site-packages\nuitka\MainControl.py", line 739, in main
    main_module = createNodeTree(filename=filename)
  File "c:\Python37-64\lib\site-packages\nuitka\MainControl.py", line 142, in createNodeTree
    Optimization.optimize(main_module.getOutputFilename())
  File "c:\Python37-64\lib\site-packages\nuitka\optimizations\Optimization.py", line 619, in optimize
    finished = makeOptimizationPass(initial_pass=False)
  File "c:\Python37-64\lib\site-packages\nuitka\optimizations\Optimization.py", line 530, in makeOptimizationPass
    changed = optimizeModule(current_module)
  File "c:\Python37-64\lib\site-packages\nuitka\optimizations\Optimization.py", line 168, in optimizeModule
    changed = optimizeCompiledPythonModule(module)
  File "c:\Python37-64\lib\site-packages\nuitka\optimizations\Optimization.py", line 127, in optimizeCompiledPythonModule
    Plugins.considerImplicitImports(module=module, signal_change=signalChange)
  File "c:\Python37-64\lib\site-packages\nuitka\plugins\Plugins.py", line 160, in considerImplicitImports
    plugin.considerImplicitImports(module, signal_change)
  File "c:\Python37-64\lib\site-packages\nuitka\plugins\PluginBase.py", line 167, in considerImplicitImports
    signal_change=signal_change,
  File "c:\Python37-64\lib\site-packages\nuitka\plugins\PluginBase.py", line 462, in recurseTo
    reason=reason,
  File "c:\Python37-64\lib\site-packages\nuitka\importing\Recursion.py", line 167, in recurseTo
    reason=reason,
  File "c:\Python37-64\lib\site-packages\nuitka\importing\Recursion.py", line 83, in _recurseTo
    is_shlib=module_kind == "shlib",
  File "c:\Python37-64\lib\site-packages\nuitka\tree\Building.py", line 924, in decideModuleTree
    source_ref=source_ref,
  File "c:\Python37-64\lib\site-packages\nuitka\nodes\ModuleNodes.py", line 264, in __init__
    setLocalsDictType(self.module_dict_name, "module_dict")
  File "c:\Python37-64\lib\site-packages\nuitka\nodes\LocalsScopes.py", line 31, in setLocalsDictType
    assert locals_dict_name not in locals_dict_handles, locals_dict_name
AssertionError: globals_pydantic$__init__
```

My installed app list with the relevant FastAPI packages:
```
python.exe -m pip list
Package                 Version
----------------------- ----------
certifi                 2019.11.28
cffi                    1.13.2
chardet                 3.0.4
Click                   7.0
colorama                0.4.3
cryptography            2.8
**fastapi                 0.48.0**
h11                     0.9.0
idna                    2.8
multiprocessing-logging 0.3.0
Nuitka                  0.6.8rc2
pip                     20.0.2
progress                1.5
psutil                  5.6.7
pyarmor                 5.8.6
pycparser               2.19
pycryptodomex           3.9.4
**pydantic                1.4**
pyminifier              2.2
pyOpenSSL               19.1.0
python-tr               0.1.2
pywin32                 227
requests                2.22.0
requests-pkcs12         1.6
requests-toolbelt       0.9.1
setuptools              41.2.0
six                     1.13.0
**starlette               0.12.9**
tblib                   1.6.0
urllib3                 1.25.7
**uvicorn                 0.11.2**
vboxapi                 1.0
websockets              8.1
WMI                     1.4.9
```

I thought of a second SSCEE which would be simpler and triggers the same result

```
import pydantic

print('hello world')
```

When compiled with `python.exe -m nuitka pydantic.py` I get the same error.

Hope this might help identifying what's going on.
Best regards.

**Comment by kayhayen** [comment]
Just as an side, don't file against factory version, it's supposed to be known broken many times. But that's not an issue right now. There is nothing problematic on it right now.

Yes, definitely a core bug. I wasn't able to get your first reproducer to not give another error (an `ImportError` for `routing`) yet, but I was on Linux then. I wondered if it is related to the filename and casing on Windows, where two modules can relatively well collide for names.

**Comment by kayhayen** [comment]
So, seems like the pydantic modules gets loaded twice, clearly something that should never happen:

> <Node 'COMPILED_PYTHON_MODULE' with {'filename': 'C:\\Python37_64\\lib\\site-packages\\pydantic\\__init__.py', 'module_name': <ModuleName pydantic.__init__>}>
> <Node 'COMPILED_PYTHON_MODULE' with {'filename': 'C:\\Python37_64\\lib\\site-packages\\pydantic\\__init__.py', 'module_name': <ModuleName pydantic.__init__>}>

I was expecting paths to be different or something, but clearly not the case. Something really smelly is happening here. Still suspicious why this only happens on the case sensitive system.

**Comment by deajan** [comment]
So this might not be a Nuitka problem but a pydantic one ?
Should I open an issue there?

**Comment by kayhayen** [comment]
No doubt a Nuitka problem, I just don't know which one yet. There is an internal caching mechanism that ought to prevent loading already loaded modules again, that must fail, and it's all our fault surely.

**Comment by kayhayen** [comment]
Ok, so for reasons that most probably had to do with inabilities regarding unicode paths, Nuitka is using relative paths to reference modules, to key them in the import cache. Also the import cache handles the `__init__.py` by removing it, where as when checking it, it's not removed. This allows for the double load of the Python package there.

Making the paths unique, so they serve as a key, makes sense, but having two places with an algorithm for how to do it, not a good idea. I am unsure, why this doesn't happen more often, it apparently needs a module that becomes unused and then used again, but pydantic wouldn't be that special there. But it's also not a `CompiledPythonPackage`, but only a module there, so this is maybe why this becomes visible, and that would be another bug.

**Comment by kayhayen** [comment]
Ok, we defined an implicit dependency do `pydantic.__init__` for `pydantic` by mistake. And this then escapes the caching mechanism. That would include the code of `pydantic` twice, once as this module, and then as the proper package, only the top level, no big harm therefore, and the module is very probably very unused, and remains an issue. I suspect this might be some form of buggy output of hinted compilation for packages, but that's only a guess.

**Comment by kayhayen** [comment]
I remain puzzled why this only happens on Windows still. It is not special in any other way than that relpath is less often usable, but it falls back to abspath there, when two disks are used, but that's not what I am doing.

**Comment by kayhayen** [comment]
My `ImportError` for `fastapi.routing` is what I get immediately after. @JorjMcKie mentioned seeing something similar with new pandas. Nuitka claims to have it included, but it fails at import time, this seems serious too. That has also happened on Linux, and apparently is cross platform problem. The importing code seemed very robust, but alas ... there is never a last bug apparently.

**Comment by JorjMcKie** [comment]
After some trials it seems to me now that there is a pattern like this:

1. `pandas.core.groupby.generic` imports `pandas.core.frame`
2. `pandas.core.frame` then imports in turn `pandas.core.groupby.generic`

When it comes to point 2, the import fails allthough both modules are part of the binary (confirmed via ``--show-modules``).

**Comment by kayhayen** [comment]
@JorjMcKie It seems there is a bug of Nuitka, that I hope to find shortly. The relative import is done with the proper level 1 as `from .routing import something`, but I have no evidence of the meta path based importer being asked about it at all. So I am getting this:

> ImportError: cannot import name 'routing' from 'fastapi' (C:\Users\kayha\repos\Py2C\MINILD~1.DIS\fastapi\__init__.py)

after these:

> import fastapi.utils # considering responsibility
> import fastapi.utils # claimed responsibility (compiled)
> Loading fastapi.utils
> import dataclasses # considering responsibility
> import dataclasses # claimed responsibility (compiled)
> Loading dataclasses

What normally always happens is that you get a trace that says "denied responsibility", definintely a "considering responsibility" for all modules, it gets asked about. But it does not appear to happen.

Maybe the common denominator is something messing with the meta path at run time, or starting to claim responsiblity of everything and having placed itself before Nuitka's, must be. Otherwise I cannot explain the lack of traces from  `--python-flag=-v`. I will enable import tracing now, just to see if the `__import___` call is to blame (but Nuitka must be asked about non-found modules for sure).

**Comment by kayhayen** [comment]
What you describe, @JorjMcKie is a cyclic dependency then?

**Comment by JorjMcKie** [comment]
> What you describe, @JorjMcKie is a cyclic dependency then?

yes, look at this (dist folder name abbreviated for better readability):

```
$ pandas-test
Traceback (most recent call last):
  File "$DIST\pandas-test.py", line 2, in <module>
  File "$DIST\pandas\__init__.py", line 54, in <module pandas>
  File "$DIST\pandas\core\api.py", line 29, in <module pandas.core.api>
  File "$DIST\pandas\core\groupby\__init__.py", line 1, in <module pandas.core.groupby>
  File "$DIST\pandas\core\groupby\generic.py", line 60, in <module pandas.core.groupby.generic>
  File "$DIST\pandas\core\frame.py", line 100, in <module pandas.core.frame>
ImportError: cannot import name 'generic' from 'pandas.core.groupby' ($DIST\pandas\core\groupby\__init__.py)
```

(Pandas seems to always fully quality where it imports stuff from, i.e. not using "." or ".." references.)

`pandas.core.groupby.generic` in line 60 imports something from `pandas.core.frame`, which gets imported for the first time at this point. This guy then tries to import (some other) things from `pandas.core.groupby.generic` again in its line 100.

**Comment by kayhayen** [comment]
So, this should be something we can make a simple reproducer out of. I believe the `__spec__` has a `_initializing` field, that is set to `True` during module execution in Nuitka), to `False` only after the module finished. This could cause it to not exist when the cylic import happens.

I need to research, but I image the attribute does not get set that way precisely in CPython. So this would be a minor compatibility bug then with large effect.

**Comment by kayhayen** [comment]
I see this function on the call stack doing the diff, and I think CPython goes into the first branch:
```

def _load_unlocked(spec):
    # A helper for direct use by the import system.
    if spec.loader is not None:
        # not a namespace package
        if not hasattr(spec.loader, 'exec_module'):
            return _load_backward_compatible(spec)

    module = module_from_spec(spec)
    with _installed_safely(module):
        if spec.loader is None:
            if spec.submodule_search_locations is None:
                raise ImportError('missing loader', name=spec.name)
            # A namespace package so do nothing.
        else:
            spec.loader.exec_module(module)

    # We don't ensure that the import-related module attributes get
    # set in the sys.modules replacement case.  Such modules are on
    # their own.
    return sys.modules[spec.name]
```

The Nuitka loader has no `exec_module`, but Python 3.7 might have them, from what I see. This will be possible to cause such control flow differences, and Nuitka emulating `_initializing` in either case, on its own, may not be a good idea and even become unnecessary if we let importlib do its spiel here.

So, adding `exec_module` will be a kind of stunt, but improves compatibility, plus I need to check which Python version should have it and which not. I am pretty sure at 3.4/5 things were not that way yet.

**Comment by kayhayen** [comment]
Nah, all of Python3 from 3.4 definitely did it this way, so basically the since the `__spec__` was invented.

But I think at one point, they started to allow these recursive imports, that would give errors. In fact, I vaguely remember dealing with similar stuff before. Maybe it's just that the solution is not working anymore in latest versions, anyway, this is possible to get under control. Just need to implement `exec_module` and `create_module` for Nuitka's loader it seems.

**Comment by kayhayen** [comment]
Funny side not, for plugins we have code like this:

```
def importFilePy3NewWay(filename):
    """ Import a file for Python versions 3.5+.
    """
    import importlib.util  # pylint: disable=I0021,import-error,no-name-in-module

    spec = importlib.util.spec_from_file_location(filename, filename)
    user_plugin_module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(user_plugin_module)
    return user_plugin_module
```

That ends up being used for user plugins and probably works. The `SourceFileLoader` must have seen changes that disallow the use. I guess, if we make a test, we would find that 3.5 is the cut off for this being done for source modules, which then allowed the cyclic dependency to work so that new code increasingly is not worried in using it, and we find more and more cases.

**Comment by JorjMcKie** [comment]
> I am pretty sure at 3.4/5 things were not that way yet.

When I mentioned Linux having the same thing happening, it was Python 3.6 which I meant ...

**Comment by kayhayen** [comment]
The 2 step initialization process, is kind of a bigger change. Module gets created first, and then executed. Due to being created first, I assume, it's sooner in `sys.modules` and therefore usable for the import statements during its execution.

The preload and postload code of course need to still be run at appropiate times, and this will take some re-factoring, and of course, the module object creation becomes then version dependent and name dependent, since `__main__` works differently of course (not loaded by import). But I might want to split this off for older Python as well, simply because it saves boiler plate C code, then it's probably not as bad. Let's see if I get something working soonish.

**Comment by kayhayen** [comment]
But this theory took a hit when I saw this in module code very early on, before any code runs:

```
// Seems to work for Python2.7 out of the box, but for Python3, the module
// doesn't automatically enter "sys.modules", so do it manually.
#if PYTHON_VERSION >= 300
    {
        int r = PyDict_SetItemString(PyImport_GetModuleDict(), module_full_name, module_%(module_identifier)s);
        assert(r != -1);
    }
#endif
```

So this has to be about something else then, we will see. That of course is among the boiler plate code that can be moved to our static C code then.

**Comment by kayhayen** [comment]
As an side, the standalone import hinting is giving very different outputs that the pure Python one, mostly because of `$PYTHONPATH` replacements not done with the dist folder. I am not sure, but I guess, we ought to use a similar replacement for the binary directory, or maybe adding it to sys.path before tracing will do. I found it surpringly useless in its current form for comparative standalone diagnosis. It was the first time I tried to use it for that, mostly because I was initially expecting a import argument deviation.

But basically @JorjMcKie how about we get there to be able to run the program compiled and uncompiled, in standalone, and see the first difference reported. There are a couple of imports, only Nuitka does, e.g. `os` early on, I think for path name manipulations, or maybe `importlib`, these are tricky too. But a meaningful diff would be a valuable debug tool. Then on top of that, I think there is a way to trace failed DLL loads, is there? If we traced those as well, even if only on Windows, we would be good. But for Linux, with strace, I would nearly know how to get there too.

**Comment by kayhayen** [comment]
I just noted, that trigger modules for pre- and postload must be compiled, or they might not get used, in case they were included as bytecode (which often makes a lot of sense though), they would be skipped. I am removing the special handling, but making those go through the same handling recursively.

**Comment by kayhayen** [comment]
Ok, this is very strange. I am very nearly sure, this cannot have happened, but the `asyncio` package, that is included as bytecode started to fail with relative imports in non-package. Nothing was setting `__package__`, right. But this ought to not have worked before. I was very confused by that. The fix is obviously to set it for `__packages__` because `PyModule_New` won't do it. But why am I seeing this now?!

**Comment by kayhayen** [comment]
And in relatively bad news, the 2 step loading doesn't change the error for recursion. Back to square one?

**Comment by kayhayen** [comment]
My line of attack for this is for now to make the import tracing properly support standalone programs, and to minimize the differences to normal Python. I am absolutely uncertain where and why it happens that the cycle leads to this kind of error. I thought we make sure it's in `sys.modules` and that ought to be enough. But apparently that is not the case.

**Comment by kayhayen** [comment]
I succeeded at making the hint tracing more compatible with standalone mode. So far I had only used for some standard programs and modules, to see where levels, etc. default to, and how details must be handled for `__import__`.

In the course of it, I enhanced Nuitka to only use the FrozenImporter at the very start, and then to remove its other frozen modules, and to include them as the normal bytecode loader, which is more compatible. I didn't figure out why, but FrozenImporter uses the current directory in its tracebacks, which is weird. So I just limited it, and then added current directory to be considered PYTHONPATH for import tracing when running compiled in standalone.

**Comment by kayhayen** [comment]
So, of course from `.fastapi import routing` should actually consider the module `fastapi.routing` but the error says `cannot import name 'routing' from 'fastapi'` which is strange. There is code in Nuitka that does this:

```
        if level and python_version >= 350:
            emit(
                """\
if (PyModule_Check(%(from_arg_name)s)) {
   %(to_name)s = IMPORT_NAME_OR_MODULE(
        %(from_arg_name)s,
        (PyObject *)moduledict_%(module_identifier)s,
        %(import_name)s,
        %(import_level)s
    );
} else {
   %(to_name)s = IMPORT_NAME(%(from_arg_name)s, %(import_name)s);
}
"""
````

I am not sure, why level is checked for non-zero here. The absolute imports of Python3 ought to also do this, so there would be no point. Probably when I first came across this, it was that way, so this can probably be safely removed.

**Comment by kayhayen** [comment]
Also, my goodness, I am bad at this:

```
            PyObject *fromlist = PyTuple_New(1);
            PyTuple_SetItem(fromlist, 0, import_name);

            result = IMPORT_MODULE5(const_str_empty, globals, globals, fromlist, level);

            Py_DECREF(fromlist);
```

I think this just looses a reference to the imported name, which probably looses a reference. I need to add a test to prove this, but definitely a source of corruption. The `PyTuple_SetItem` steals a reference only, and we were none given in that code. :/

**Comment by kayhayen** [comment]
So, for level 0, different code is needed, where the __name__ is built from the original module, the level 1 is entirely different. I think I will be able to add this shortly. Woohoo. :)

**Comment by kayhayen** [comment]
I just pushed code that ought to work with this to factory, but right now I am on the road, unable to test with our case here.

**Comment by kayhayen** [comment]
It seems to work, but it reveals more hidden dependencies of `uvicorn` once it runs, e.g. `uvicorn.logging` is dynamically imported by `uvicorn.config` which at first fooled me into thinking that something had regressed.

**Comment by kayhayen** [comment]
So, on factory there is a solution for this. The extract that I made, refuses to run, missing fapi_test or so, but that is the same for Python and Nuitka, so probably just some incompleteness.

I added a missing dependency, removed a wrong one for pydantic, and most importantly, I added support for absolute imports to also be able to import modules. This allows for cyclic imports of modules, because the module needs no name assigned. 

In case a package imported its child and was run first, this never became visible, as the mere attribute lookup was good enough.

Very complicated stuff, and like the root cause of a couple of issues we recently saw, @JorjMcKie . I am going to ignore these, and ask people to re-test once I did a pre-release, which should be soon now.

**Comment by kayhayen** [comment]
This is now on develop, and going to be part of the next release.

**Comment by deajan** [comment]
I can confirm that pydantic / fapi now compiles with Nuitka 0.6.8rc4.

As you suggested, uvicorn ASGI server won't run, but that wasn't your target for the issue.
Maybe as a side effect of https://github.com/Nuitka/Nuitka/issues/613, the error message stating that it cannot run ASGI app has some unicode signs in it:

![image](https://user-images.githubusercontent.com/4681318/75539112-03587600-5a1a-11ea-8eba-d38a56ef4ed0.png)

**Comment by deajan** [comment]
Just for the records, when using hinted compilation, fastapi + uvicorn works great with Nuitka 0.6.8rc5, since we catch the dynamic imports :)

Closing this issue since pydantic now compiles with circular imports.

**Comment by kayhayen** [comment]
The release 0.6.8 was just made and contains the correction.

**Comment by Leezj9671** [comment]
I got `ERROR:    Error loading ASGI app. Could not import module "fapi_test".`  when used the demo.

Python 3.7.10 FastAPI Uvicorn Nuitka all the newest version on Ubuntu 18.04

**Comment by kayhayen** [comment]
@Leezj9671 make a new issue
